hypothesis-python/tests/nocover/test_stateful.py

```diff
@@ -266,3 +285,100 @@ def bad_invariant_d(self):
+
+    @rule(target=a, a=st.integers())
+    def add_a(self, a):
+        self.calls.add("add")
+        return a
+
+    @rule(v=a)
+    def f(self, v):
+        self.calls.add("f")
+
+    @precondition(lambda self: "add" in self.calls)
+    @rule(value=st.integers())
+    def unrelated(self, value):
+        self.calls.add("unrelated")
+
+    @rule()
+    def invariant(self):
+        # force all three calls to be made in a particular order (with the
+        # `unrelated` precondition) so we always shrink to a particular counterexample.
+        assert len(self.calls) != 3
+
+
+def test_unrelated_rule_does_not_use_var_reference_repr():
+    # we are specifically looking for state.unrelated(value=0) not being replaced
+    # with state.unrelated(value=a_0). The `unrelated` rule is drawing from
+    # st.integers, not a bundle, so the values should not be conflated even if
+    # they're both 0.
+    assert_runs_to_output(
+        UnrelatedCall,
+        """
+        state = UnrelatedCall()
+        a_0 = state.add_a(a=0)
+        state.f(v=a_0)
+        state.unrelated(value=0)
+        state.invariant()
+        state.teardown()
+        """,
+    )
+
+
+class SourceSameAsTarget(RuleBasedStateMachine):
+    values = Bundle("values")
+
+    @rule(target=values, value=st.lists(values))
+    def f(self, value):
+        assert len(value) == 0
+        return value
+
+
+class SourceSameAsTargetUnclearOrigin(RuleBasedStateMachine):
+    values = Bundle("values")
+
+    def __init__(self):
+        super().__init__()
+        self.called = False
+
+    @rule(target=values, value=st.just([]) | st.lists(values))
+    def f(self, value):
+        assert len(value) == 0 or not self.called
+        # ensure we get two calls: one from st.just([]) as the simplest initial
+        # example, then the second one is the failure from st.lists(values) of
+        # [[]].
+        self.called = True
+        return value
+
+
+@pytest.mark.parametrize(
+    "TestClass", [SourceSameAsTarget, SourceSameAsTargetUnclearOrigin]
+)
+def test_source_same_as_target_doesnt_replace_in_output(TestClass):
+    # one might think that we could track strategies that draw from a bundle and
+    # replace any references to that value with the var reference repr (like values_0).
+    # SourceSameAsTarget shows a case where this might seem desirable.
+    #
+    # The problem is that the strategy may be arbitrarily complicated and use a value
+    # with the same repr as values_0, but from a different source.
+    # SourceSameAsTargetUnclearOrigin shows that this can happen - st.lists(values)
+    # has chosen the value [] from the bundle as a list element, but this could
+    # have also been generated via st.just([]). What we do *not* want to display
+    # here is state.f(value=[values_0]).
+    assert_runs_to_output(
+        TestClass,
+        f"""
+        state = {TestClass.__name__}()
+        values_0 = state.f(value=[])
+        state.f(value=[[]])
```

Comment:
I think here we actually _do_ want 
```suggestion
        values_0 = state.f(value=[])
        state.f(value=[values_0])
```
because e.g. lists are mutable and so it's valuable to show the variable name which communicates the identity of the value.

I realize that this is Quite A Can Of Worms, what with not wanting `values_0 = state.f(value=values_0)`; your motivating case in the changelog is spot-on but here and in the changed test I think the change is probably for the worse :/

Reply:
agree that this case is worse than it used to be, but it's also not a correct transformation to make in general; see `SourceSameAsTargetUnclearOrigin`, where we would get `values_0 = state.f(value=[values_0])` even if the inner empty list came from `st.just([])`. And one could imagine more complicated cases of building up the inner list via some alternative st.composite where this would be more misleading.

I think we have to decide which of the two evils is worse...I believe the only case worsened by this is strategies involving bundles, which isn't great but is less common than rules involving bundles. The latter case collides relatively frequently thanks to shrinking normalizing to e.g. lots of zeros.